File: tests/xcalablemp/global-view/intrinsic/C/intrinsic_gather2x2.c
#include <stdio.h>
#include <stdlib.h>
#include "xmp.h"
extern void xmp_gather(void *x_d, void *a_d, ... );
#pragma xmp nodes p(2,2)
#pragma xmp template t(0:3,0:3)
#pragma xmp distribute t(block,block) onto p 
  int x[4][4],a[4][4],idx0[4][4],idx1[4][4];
#pragma xmp align a[i][j] with t(i,j) 
#pragma xmp align idx0[i][j] with t(i,j) 
#pragma xmp align idx1[i][j] with t(i,j) 
#pragma xmp align x[i][j] with t(i,j) 

int main()
{
  int i,j;

#pragma xmp loop (i,j) on t(i,j)
  for(i=0;i<4;i++)
    for(j=0;j<4;j++)
      a[i][j]=i*4+j;

#pragma xmp loop (i,j) on t(i,j)
  for(i=0;i<4;i++)
    for(j=0;j<4;j++)
      {
	idx0[i][j]=i;
	idx1[i][j]=j;
      }

#pragma xmp loop (i,j) on t(i,j)
  for(i=0;i<4;i++)
    for(j=0;j<4;j++)
      x[i][j]=0;

  xmp_gather(xmp_desc_of(x),xmp_desc_of(a),xmp_desc_of(idx0),xmp_desc_of(idx1));


  int result = 0;
#pragma xmp loop (i,j) on t(i,j)
  for(i=0;i<4;i++)
    for(j=0;j<4;j++)
      if(x[i][j]!=a[i][j])
        result =-1;

#pragma xmp reduction(+:result)

#pragma xmp task on p(1,1)
  {
    if (result == 0){
      printf("PASS\n");
    }
    else{
      printf("ERROR\n");
      exit(1);
    }
  }

  return 0;
}

File: tests/XMP/global-view/intrinsic/F/intrinsic_gather2x2.f90
program main
  !$xmp nodes p(2,2)
  !$xmp template t(4,4)
  !$xmp distribute t(block,block) onto p 
  integer(8):: a(4,4),x(4,4),idx1(4,4),idx2(4,4)
  !$xmp align a(i,j) with t(i,j)
  !$xmp align x(i,j) with t(i,j)
  !$xmp align idx1(i,j) with t(i,j)
  !$xmp align idx2(i,j) with t(i,j)
  integer:: i,j,answer = 0

  !$xmp loop on t(i,j)
  do i=1,4
     do j=1,4
        a(i,j)=i*4+j
     end do
  end do

  !$xmp loop on t(i,j)
  do i=1,4
     do j=1,4
        x(i,j)=0
     end do
  end do

  !$xmp loop on t(i,j)
  do i=1,4
     do j=1,4
        idx1(i,j)=i
        idx2(i,j)=j
     end do
  end do

  call xmp_gather(xmp_desc_of(x),xmp_desc_of(a),xmp_desc_of(idx1),xmp_desc_of(idx2))

  !$xmp loop on t(i,j)
  do i=1,4
     do j=1,4
        if(x(i,j).ne.a(i,j)) then
           answer = -1
        endif
     end do
  end do

  !$xmp reduction(+:answer)

  !$xmp task on p(1,1)
  if ( answer /= 0 ) then
     write(*,*) "ERROR"
     call exit(1)
  endif

  write(*,*) "PASS"
  !$xmp end task


end program main
